### DISCUSSION TITLE
STM32G030 with u8g2 library problem

### DISCUSSION THREAD

**Discussion by sunnyguhz** [stm32duino/Arduino_Core_STM32#2116] [Libraries] [answered]
**Describe the bug**
Hi I have an ST7567 LCD panel that has a 4-pin SPI connection, and I use U8g2 library.

The panel works fine with Arduino(ESP8266 or STM32F103C8T6) under Hardware and Software SPI mode. 

It displays okay in Software SPI mode with STM32G030.
```
#include <U8g2lib.h>
#include <SPI.h>

//U8G2_ST7567_OS12864_1_4W_HW_SPI u8g2(U8G2_R2,  /* cs=*/ PA4, /* dc=*/ PA6, /* reset=*/ PB1);  
U8G2_ST7567_OS12864_1_4W_SW_SPI u8g2(U8G2_R2, /* clock=*/PA1, /* data=*/ PA2, /* cs=*/ PA4, /* dc=*/ PA6, /* reset=*/ PB1); 

void setup(void) {
  u8g2.begin(); 
}

void loop(void) {
  u8g2.firstPage();
  do {
    u8g2.setFont(u8g2_font_ncenB10_tr);
    u8g2.drawStr(0,24,"Hello World!");
  } while ( u8g2.nextPage() );
  delay(1000);
}
```

But it it just won't show anything in hardware SPI mode.
```
#include <U8g2lib.h>
#include <SPI.h>

U8G2_ST7567_OS12864_1_4W_HW_SPI u8g2(U8G2_R2,  /* cs=*/ PA4, /* dc=*/ PA6, /* reset=*/ PB1);  
//U8G2_ST7567_OS12864_1_4W_SW_SPI u8g2(U8G2_R2, /* clock=*/PA1, /* data=*/ PA2, /* cs=*/ PA4, /* dc=*/ PA6, /* reset=*/ PB1); 

void setup(void) {
  u8g2.begin(); 
}

void loop(void) {
  u8g2.firstPage();
  do {
    u8g2.setFont(u8g2_font_ncenB10_tr);
    u8g2.drawStr(0,24,"Hello World!");
  } while ( u8g2.nextPage() );
  delay(1000);
}
```

**Desktop (please complete the following information):**
 - OS: [ Windows10]
 - Arduino IDE version: [2.2.0]
 - STM32 core version: [2.6.0]
 - Tools menu settings if not the default: [Debug:None, Optimize:Smallest(-Os) with LTO, USB:None, U(S)ART:Disable]
 - Upload method: [SWD]

**Board (please complete the following information):**
 - Name: [Generic G030C8Tx]
 - Hardware Revision: [I don't know]

**Comment by fpistm** [comment, marked as answer]
By default generic G030C8 uses these pins for SPI instance: https://github.com/stm32duino/Arduino_Core_STM32/blob/01e4f0c9a8d20fc20125fd5be539a1ef3bf9d546/variants/STM32G0xx/G030C(6-8)T/variant_generic.h#L101-L109
Do you use them?
Else you can change them using setxxx api:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/API#change-default-spi-instance-pins

**Reply by sunnyguhz** [reply]
Yes，I used these pins. but it don't work

**Reply by fpistm** [reply]
Unfortunately, I will not be able to help on this as I do not have this hardware. Check your wiring and pin assignment.
For example the generic use for default instance of SPI and Wire PA2. Anyway, it seems you disabled the Serial, so it should not be an issue.
Also if your board have an HSE then you can redefine the [systemCoreClock_config](https://github.com/stm32duino/Arduino_Core_STM32/wiki/Custom-definitions#systemclock_config).
Finally, I would open an issue on U8G2 repo.

**Reply by sunnyguhz** [reply]
This is already the best way. Think you very much.

**Reply by sunnyguhz** [reply]
Hi, I have found the cause of this problem. 
My chinese net friend tell me that MISO cannot be occupied even if MISO not need for LCD panel. And it all so can be change to another pins like PA11 or PB4.

I think it is a bug for hardware SPI mode with STM32G030. Can you do something for problem?

**Reply by fpistm** [reply]
No. As SPI works fine on G0. Don't know what is your issue but not  link to the core.

**Reply by sunnyguhz** [reply]
Obviously, SPI is working with a bug on G0.

**Reply by fpistm** [reply]
Not agree with your analyze. My guess is you have pin wiring issue or conflict. Or wrong library conf.
